### PR TITLE
OSX migration

### LINKED ISSUE #149: replace ld_logger by scan-build.py (intercept)
Examine the possibility of replacing ld-logger by lib ear:https://github.com/rizsotto/Bear
Bear works on OSX and tested & used by a larger community.

### PR DESCRIPTION
**Pull request by tmsblgh**
- Add scan-build.py (intercept) support, used instead of ld_logger on OS X
- Update travis.yml
- Using Makefiles for testing
- Expand README with OS X things
- Resolves #149

### CODE REVIEW

File: tests/test_env_setup.sh [hunk 1]
```diff
@@ -2,17 +2,27 @@
 
 if [[ "$OSTYPE" == "linux-gnu" ]]; then
```
Comment (gyorb) [inline tests/test_env_setup.sh]: Could we merge these two if statements for setting TEST_TESTS_DIR and TEST_CODECHECKER_PACKAGE_DIR environment variables?
If OSTYPE is not set the default value should be linux-gnu.

**Comment by gyorb** [comment]
More descriptive comment messages would be better instead of "Small changes..." :wink:

**Comment by r0mai** [comment]
Official OS X support would be very welcome. I'm in the process of trying to make this branch work on my machine. I'll report the issues I encounter.

File: codechecker_lib/build_manager.py [hunk 2]
```diff
@@ -37,11 +45,22 @@ def perform_build_command(logfile, command, context, silent=False):
...
+
+    # Run user's commands in shell
+    else:
+        if platform.system() == 'Linux':
+            LOG.debug_analyzer("with ld logger ...")
+            log_env = analyzer_env.get_log_env(logfile, context, original_env)
+            if 'CC_LOGGER_GCC_LIKE' not in log_env:
+                log_env['CC_LOGGER_GCC_LIKE'] = 'gcc:g++:clang:clang++:cc:c++'
 
     LOG.debug_analyzer(log_env)
```
Comment (r0mai) [inline codechecker_lib/build_manager.py]: I got an ugly python error if we get here and check_intercept returned false and we're not on Linux.

```
[ERROR] - Failed to run: intercept-build"
[ERROR] - local variable 'log_env' referenced before assignment
Traceback (most recent call last):
  File "/Users/r0mai/codechecker_package/CodeChecker/cc_lib/python2.7/codechecker_lib/arg_handler.py", line 226, in handle_check
    context)
  File "/Users/r0mai/codechecker_package/CodeChecker/cc_lib/python2.7/codechecker_lib/build_manager.py", line 167, in generate_log_file
    silent=silent)
  File "/Users/r0mai/codechecker_package/CodeChecker/cc_lib/python2.7/codechecker_lib/build_manager.py", line 65, in perform_build_command
    LOG.debug_analyzer(log_env)
UnboundLocalError: local variable 'log_env' referenced before assignment
```

A more user friendly error message would be nice :)

Comment (tmsblgh) [inline codechecker_lib/build_manager.py]: Hi, I forgot about this option. I've made some changes, you can check it. Thanks for your help :) @r0mai

**Comment by gyorb** [comment]
Do you still experience some issues @r0mai ?

**Comment by r0mai** [comment]
I didn't have much time as I expected, but here where I'm now:

I followed the documentation more or less blindly https://github.com/tmsblgh/codechecker-osx-migration/blob/7ed507526723fab6bde1b601bf6f6825da416d9f/README.md#os-x

At the `./build_package.py -o ~/codechecker_package` line I get the following output:

```
$ ./build_package.py -o ~/codechecker_package
[INFO] - Checking source: thrift
[INFO] - Done.
[INFO] - Checking source: codemirror
[INFO] - Done.
[INFO] - Checking source: jsplumb
[INFO] - Done.
[INFO] - Checking source: marked
[INFO] - Done.
[INFO] - Checking source: dojotoolkit
[INFO] - Done.
[INFO] - Checking source: highlightjs
[INFO] - Done.
[INFO] - Checking source: highlightjs_css
[INFO] - Done.
[INFO] - Getting external dependecies done.
[INFO] - Removing previous package
[INFO] - Creating package layout
[INFO] - Checking source: llvm scan-build-py (intercept)
[ERROR] - Not exists, build ld logger in Linux only
[INFO] - Generating thrift files ...
[INFO] - Generating documentation ...
[INFO] - Creating package finished successfully.
```

Later in the documentation it is mentioned, that codecheker has to be able to find `intercept-build` or `scan-build`, so I tried adding it to the path:

```
$ PATH="$PATH:$HOME/build/clang-for-codechecker/llvm/tools/clang/tools/scan-build-py/bin" ./build_package.py -o ~/codechecker_package
[INFO] - Checking source: thrift
[INFO] - Done.
[INFO] - Checking source: codemirror
[INFO] - Done.
[INFO] - Checking source: jsplumb
[INFO] - Done.
[INFO] - Checking source: marked
[INFO] - Done.
[INFO] - Checking source: dojotoolkit
[INFO] - Done.
[INFO] - Checking source: highlightjs
[INFO] - Done.
[INFO] - Checking source: highlightjs_css
[INFO] - Done.
[INFO] - Getting external dependecies done.
[INFO] - Removing previous package
[INFO] - Creating package layout
[INFO] - Checking source: llvm scan-build-py (intercept)
[INFO] - Available
[INFO] - Generating thrift files ...
[INFO] - Generating documentation ...
[INFO] - Creating package finished successfully.
```

So first thing to note, that the documentation here could be improved by adding this PATH requirement to the install & setup section.

After that I tried it on a test project with a pretty obvious bug: ([full gist](https://gist.github.com/r0mai/39e8a1fab3f46d4c5589f1449d21b83b))

```cpp
int main() {
	int a[11];
	a[123] = 15;
}
```

```
$ mkdir build && cmake ..
[...]
$ CodeChecker check -n test_project -b "make"
[INFO] - Starting build ...
Scanning dependencies of target cc-test
[ 50%] Building CXX object CMakeFiles/cc-test.dir/main.cpp.o
/Users/r0mai/prezi/codechecker-test/main.cpp:3:2: warning: array index 123 is past the end of the array (which contains 11 elements) [-Warray-bounds]
        a[123] = 15;
        ^ ~~~
/Users/r0mai/prezi/codechecker-test/main.cpp:2:2: note: array 'a' declared here
        int a[11];
        ^
1 warning generated.
[100%] Linking CXX executable cc-test
[100%] Built target cc-test
[INFO] - Build finished successfully.
[INFO] - Waiting for check results on [49733]
[INFO] - Static analysis is starting ...
[INFO] - clangsa analyzed main.cpp successfully.
[INFO] - clang-tidy analyzed main.cpp successfully.
[INFO] - ----==== Summary ====----
[INFO] - Total compilation commands: 2
[INFO] - Successfully analyzed
[INFO] -   clangsa: 1
[INFO] -   clang-tidy: 1
[INFO] - ----=================----
[INFO] - Analysis length: 0.315434932709 sec.
[INFO] - Analysis has finished.
[INFO] - To view results run:
CodeChecker server -w /Users/r0mai/.codechecker
```

Starting the server and opening the localhost address opens a website with a single run. But clicking on that run shows no results, nothing:

screen shot 2016-08-28 at 22 49 26

Maybe this simple program doesn't trigger any checker and that's why there are no issues? Here I'd suggest adding to the repo a simple example cpp file, build scripts and some instructions to the readme on how to check it right after the installation instructions, so the user can easily make sure that the whole infrastructure works before analyzing their big project.

I'm here now. I'll report back if I find out other stuff.

Also, I talked with @tmsblgh in private about having installer scripts which can take care of the whole installation process easily without all this manual hassle (like cloning and building a clang locally, setting up environment variables and setting files to point to that private clang build, etc).

**Comment by gyorb** [comment]
Can you view the results in the command line? ```CodeChecker cmd results -p 8001 -n test_project```

**Comment by r0mai** [comment]
No:
```
$ CodeChecker cmd results -p 8001 -n test_project
-------------------------------
File | Checker | Severity | Msg
-------------------------------
-------------------------------
```

I'll try to run it on something bigger project tonight. Can you give me some example cpp which should surely produce some report?

**Comment by gyorb** [comment]
We run codechecker multiple times on bzip2, xerces, vim.
Please try to use the newer version #363 it is updated for the latest codechecker.

You could try to enable debug mode for the analysis and for the server viewing the results by setting the environment variable ```export CODECHECKER_VERBOSE=debug``` in the shell.
Let me know if there are any error/warning messages.

**Pull request merged**